### ISSUE TITLE
Show errors in browser like Elm Reactor

### ISSUE THREAD

**Issue by TheOddler** [wking-io/elm-live#154]
It would be nice if elm build errors would be shown in the browser like elm reactor does. Or did I miss functionality that's already present?

**Comment by wking-io** [comment]
That is a good idea! I will look into what it would take to do this. If you have any ideas feel free to comment or add a PR!

**Comment by harrysarson** [comment]
I did some research into how `elm reactor` does this:

### Steps `elm reactor` takes

1. Create a server (function `run` in https://github.com/elm/compiler/blob/master/terminal/src/Develop.hs)
2. serve files
    1. serve elm files. **If there is an error call function `makePageHtml` in https://github.com/elm/compiler/blob/master/terminal/src/Develop.hs**.
    2. serve non elm files.
3. serve directory (bring up the directory listing page).
4. serve assets. **assets are defined in https://github.com/elm/compiler/blob/master/terminal/src/Develop/StaticFiles.hs**

### How `elm reactor` generates error pages

When `makePageHtml` is called it is passes (via flags) the errors generated by the compiler (as json) to the elm that runs in the webpage. `makePageHtml` generates the html for the error page. The css, images and js (compiled from elm) are defined in https://github.com/elm/compiler/tree/master/reactor.

### Possible way that `elm-live` could do this

1. Add `--report=json` as a flag to elm-make.
2. Detect when `elm make` fails - I believe this is already done.
3. Read the json into javascript rather than dumping it onto the console.
3. When `elm make` fails serve a custom html page (like that generated by `makePageHtml`) for all http requests **for html files***.
4. Pass the error json to the custom html page.
5. Directly use the files in https://github.com/elm/compiler/tree/master/reactor to create the front end.

\* The server still needs to serve the correct css, javascipt files, etc because the error html page needs to be able to fetch these to work.

ps: there may be mistakes in the above, please forgive these.

**Comment by wking-io** [comment]
I took a look at create-elm-app and they use Webpack and parse the output from the elm-loader.

I think we could make some updates to the build function to return the Elm error messages and then be able to send those to you server and have the server (elm-serve) render a custom error page using the passed elm errors.

**Comment by lucamug** [comment]
I tried to investigate the implementation of this feature too (disclaimer: I have limited knowledge of node).

I was able to export a `sendMessage` function from elm-serve to elm-live that can be used to display errors in the browser (using web sockets).

The thing that I was not able to do yet is to intercept the Elm compiler error.

Elm compiler is called in elm-live/lib/src/elm-live.js with

    const elmMake = spawnSync(args.pathToElm, ['make', ...args.elmMakeArgs], {
      stdio: [inputStream, outputStream, outputStream]
    })

using the library https://www.npmjs.com/package/cross-spawn

`inputStream` and `outputStream` are defined as `process.stdin` and `process.stdout` in `elm-live/bin/elm-live.js`

I am not sure how I can intercept those streams so that I can use them in the `sendMessage` fucntion

**Comment by lucamug** [comment]
I created a poc for showing the errors inside the browser:

* It show errors both in the browser and in the browser's console
* It show a spinner while Elm is compiling in the browser with a message "You've changed `Main.elm`. Rebuilding!"
* I removed the dependency on `supervisor`, now the web server is not reloading anymore but just the browser
* The refresh now happen only after a modification to Elm files. 
* I removed the dependency from elm-serve (to simplify the development of the poc)

If you want to try it, I uploaded the code in this branch: https://github.com/lucamug/elm-live/tree/add-elm-errors-in-browser

This is how the experience looks like:

![elm-live-testing](https://user-images.githubusercontent.com/5551094/58537085-1d67ec80-822d-11e9-9f03-192565c71bcd.gif)

**Comment by wking-io** [comment]
I have been debating for awhile if I want to remove the elm-serve dependency. My original thought was it could be a standalone server that anyone could use for elm projects, but it has been more of a pain than a benefit. I think I am going to combine the server into elm-live so that we can get this feature added ASAP! :)

**Comment by wking-io** [comment]
@lucamug I noticed some issues with the SSL. Looks like your updates always set SSL to false and when SSL is true in `elm-reload-server` the `reloadReturned` object returned from the `elm-reload` file is undefined. It is something to do with pem's callback not firing so I will help look into that, but we will need to get that fixed.

**Comment by lucamug** [comment]
@wking-io, good catch, I didn't try the ssl version. I fixed it now but I got another problem related to it: when using ssl, it launches the server in async mode using the function `pem.createCertificate`. This makes fail my way of bubbling up the `sendMessage` function producing the error `Cannot read property 'sendMessage' of undefined`. I was not able to fix this yet. I will keep looking into it, let me know if you have any suggestion.

Other notes...

I agree that with the removal of `supervisor` the package `elm-server` lost part of its appealing of workings standalone.

I added colors to errors in the browser, the browser console and the terminal (updates are in the same branch):

![elm-live-demo-2](https://user-images.githubusercontent.com/5551094/58607462-a5580000-82da-11e9-9266-75f6fb4e30f6.gif)
